src/main/java/com/aws/greengrass/lifecyclemanager/GenericExternalService.java

```diff
@@ -482,121 +498,134 @@ public boolean isPaused() {
+                            logger.atWarn("service-run-timed-out")
+                                    .log("Service failed to run within timeout, calling close in process");
+                            reportState(State.ERRORED, ComponentStatusCode.RUN_TIMEOUT);
+                            processToClose.close();
+                        } catch (IOException e) {
+                            logger.atError("service-close-error").setCause(e)
+                                    .log("Error closing service after run timed out");
+                        }
                     }
-                }
-            }, timeout, TimeUnit.SECONDS);
+                }, timeout, TimeUnit.SECONDS);
+            }
         }
     }
 
     @Override
-    protected synchronized void shutdown() {
-        logger.atInfo().log("Shutdown initiated");
+    protected void shutdown() {
+        try (LockScope ls = LockScope.lock(lock)) {
+            logger.atInfo().log("Shutdown initiated");
 
-        if (isPaused()) {
-            // Resume if paused for a graceful shutdown
-            try {
-                resume(false, false);
-            } catch (ServiceException e) {
-                // Reset tracking flag
-                paused.set(false);
-                logger.atError().setCause(e).log("Could not resume service before shutdown, process will be killed");
+            if (isPaused()) {
+                // Resume if paused for a graceful shutdown
+                try {
+                    resume(false, false);
+                } catch (ServiceException e) {
+                    // Reset tracking flag
+                    paused.set(false);
+                    logger.atError().setCause(e)
+                            .log("Could not resume service before shutdown, process will be killed");
+                }
             }
-        }
 
-        try {
-            RunResult cached = shutdownExecCache;
-            if (shouldCacheShutdownExec() && cached != null && cached.getDoExec() != null) {
-                logger.atDebug().log("Using cached shutdown command");
-                cached.getDoExec().apply();
-            } else {
-                run(Lifecycle.LIFECYCLE_SHUTDOWN_NAMESPACE_TOPIC, null, lifecycleProcesses);
-            }
-        } catch (InterruptedException ex) {
-            logger.atWarn("generic-service-shutdown").log("Thread interrupted while shutting down service");
-        } finally {
-            stopAllLifecycleProcesses();
+            try {
+                RunResult cached = shutdownExecCache;
+                if (shouldCacheShutdownExec() && cached != null && cached.getDoExec() != null) {
+                    logger.atDebug().log("Using cached shutdown command");
+                    cached.getDoExec().apply();
+                } else {
+                    run(Lifecycle.LIFECYCLE_SHUTDOWN_NAMESPACE_TOPIC, null, lifecycleProcesses);
+                }
+            } catch (InterruptedException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

**Remember**: Do this only for InterruptedException. Doing this for other types of exceptions may give rise to subtle bugs.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)